rasa/shared/core/domain.py

```diff
@@ -1074,15 +1074,109 @@ def get_active_states(self, tracker: "DialogueStateTracker") -> State:
         }
         return self._clean_state(state)
 
+    @staticmethod
+    def _remove_rule_only_features(
+        state: State,
+        rule_only_slots: Optional[List[Text]],
+        rule_only_loops: Optional[List[Text]],
+    ) -> None:
+        # remove slots which only occur in rules but not in stories
+        if rule_only_slots:
+            for slot in rule_only_slots:
+                if state.get(rasa.shared.core.constants.SLOTS, {}).get(slot):
+                    del state[rasa.shared.core.constants.SLOTS][slot]
+        # remove active loop which only occur in rules but not in stories
+        if (
+            rule_only_loops
+            and state.get(rasa.shared.core.constants.ACTIVE_LOOP, {}).get(
+                rasa.shared.core.constants.LOOP_NAME
+            )
+            in rule_only_loops
+        ):
+            del state[rasa.shared.core.constants.ACTIVE_LOOP]
+
+    @staticmethod
+    def _substitute_rule_only_user_input(state: State, last_ml_state: State) -> None:
+        if not rasa.shared.core.trackers.is_prev_action_listen_in_state(state):
+            if not last_ml_state.get(rasa.shared.core.constants.USER) and state.get(
+                rasa.shared.core.constants.USER
+            ):
+                del state[rasa.shared.core.constants.USER]
+            elif last_ml_state.get(rasa.shared.core.constants.USER):
+                state[rasa.shared.core.constants.USER] = last_ml_state[
+                    rasa.shared.core.constants.USER
+                ]
+
     def states_for_tracker_history(
```

Comment:
can we please add some tests for this in `test_domain`?

Reply:
we can, but effectively, they will duplicate the tests in `test_rule_policy.py`